Utilities/Firmware.cs

```diff
@@ -305,15 +309,9 @@ void getAPMVersion(object tempin)
 
         public string GetAPMVERSIONFile(Uri url)
         {
-            lock (urlcachelock)
-                if (!urlcacheSem.ContainsKey(url.AbsoluteUri))
-                    urlcacheSem[url.AbsoluteUri] = new SemaphoreSlim(1, 1);
-
-            urlcacheSem[url.AbsoluteUri].Wait();
-
-            try
+            using (urlcacheSem.Lock(url.AbsoluteUri))
             {
-                lock (urlcachelock)
+                lock (urlcacheSem)
                     if (urlcache.ContainsKey(url.AbsoluteUri))

```

Comment:
is this guaranteed to contain the key now? given the above using.

i understand that your the author of the library?

the intent of this function was to dedup http requests/cache and respond

Reply:
I'm not sure I follow what you mean in your first sentence.

Yes, I am the author of the AsyncKeyedLock library.

Reply:
line 
using (urlcacheSem.Lock(url.AbsoluteUri))

must create a state
that
 if (urlcache.ContainsKey(url.AbsoluteUri))
 
 will not fail on? ie is it redundant?

Reply:
No I don't think so. Imagine you fire up the application, urlcache and urlcacheSem will both be empty. If you make a new call to a URL, urlcacheSem will be used, obtains the lock and goes check urlcache to see if it's in the cache. It won't find it.